File: helpers/graph.go
package helpers

// TopologicalSortStrings returns the list of strings from a graph in topological order
// The output list is sorted from leaf to root
// NOTE: THIS DOES NOT WORK FOR CYCLIC GRAPHS!
func TopologicalSortStrings(graph map[string][]string) []string {
	out := []string{}
	handled := map[string]bool{}
	var impl func(string)
	impl = func(item string) {
		{
			_, alreadyDone := handled[item]
			if alreadyDone {
				return
			}
		}
		handled[item] = true
		for _, child := range graph[item] {
			impl(child)
		}
		out = append(out, item)
	}
	for key := range graph {
		impl(key)
	}
	return out
}

// ShortestIntPath searches for and returns the shortest path from start to end in the graph
// This uses BFS
func ShortestIntPath(graph map[int][]int, start int, end int) []int {
	seen := map[int]bool{}
	queue := [][]int{[]int{start}}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		pos := current[len(current)-1]
		seen[pos] = true
		if pos == end {
			return current
		}

		for _, neighbor := range graph[pos] {
			if !seen[neighbor] {
				queue = append(queue, append(append([]int{}, current...), neighbor))
			}
		}
	}
	return nil // No path!
}

// LongestIntPathLength returns the length of the longest path from start to somewhere in the graph
func LongestIntPathLength(graph map[int][]int, start int) int {
	seen := map[int]bool{start: true}
	steps := -1
	queue := []int{start}
	for len(queue) > 0 {
		steps++
		newQueue := []int{}
		for _, pos := range queue {
			for _, n := range graph[pos] {
				_, alreadySeen := seen[n]
				if !alreadySeen {
					newQueue = append(newQueue, n)
					seen[n] = true
				}
			}
		}
		queue = newQueue
	}
	return steps
}

File: helpers/graph.py
def topological_sort(graph):
    '''Outputs graph nodes in topological order, going from the leaf to the root
    NOTE: THIS DOES NOT WORK FOR CYCLIC GRAPHS!'''
    seen = set()
    def impl(key):
        if key in seen:
            return
        seen.add(key)
        for link in graph[key]:
            yield from impl(link)
        yield key
    for key in graph.keys():
        yield from impl(key)

def shortest_path(graph, start, end):
    '''Searches for and returns the shortest path from start to end in the graph
    Returns None on failure to find any path
    Uses BFS'''
    seen = set()
    queue = [[start]]
    while len(queue) > 0:
        current = queue.pop(0)
        pos = current[-1]
        seen.add(pos)
        if pos == end:
            return current

        for neighbor in graph[pos]:
            if neighbor not in seen:
                queue.append(current + [neighbor])
    return None

def longest_path_length(graph, start):
    '''Searches for and returns the length of the longest path from start to
    somewhere in the graph'''
    seen = {start}
    steps = -1
    queue = [start]
    while len(queue) > 0:
        steps += 1
        new_queue = []
        for pos in queue:
            for neighbor in graph[pos]:
                if neighbor in seen:
                    continue
                new_queue.append(neighbor)
                seen.add(neighbor)
        queue = new_queue
    return steps

def all_reachable(graph, start):
    '''Returns (node, distance) pairs of all reachable nodes in the graph.
    graph[node] must return a list of (neighbor, distance) pairs'''
    seen = []
    queue = [(start, 0)]
    while len(queue) > 0:
        current_node, current_dist = queue.pop(0)
        if current_node in seen:
            continue
        seen.add(current_node)

        if current_node != start:
            yield current_node, current_dist

        for neighbor_node, neighbor_dist in graph[current_node]:
            queue.append((neighbor_node, current_dist +  neighbor_dist))

        # TODO: Priority queue
        queue = sorted(queue, key=lambda n,d: d)

def dijkstra_length(graph, start, end):
    '''Returns the length of the path from start to end in the graph.
    Return -1 if no path is found.
    graph[node] must return a list of (neighbor, distance) pairs'''
    sen = set()
    queue = [(start, 0)]
    while len(queue) > 0:
        current_node, current_dist = queue.pop(0)
        if current_node == end:
            return current_dist

        if current_node in seen:
            continue
        seen.add(current_node)

        for neighbor_node, neighbor_dist in graph[current_node]:
            queue.append((neighbor_node, current_dist +  neighbor_dist))

        # TODO: Priority queue
        queue = sorted(queue, key=lambda n,d: d)

    return -1
